
### Description of columns in the "All_Info" sheet

| Column Name | Type | Description |
|-------|------|-------------|
| primer_index | string | Primer index for the construct. |
| Status | string | Indicates if obtaining a valid clone for the construct was "Successful" or "Failed". |
| Reference | string | Name of the reference. |
| n_frag | int | Number of fragments. |
| Reads_sorted_after_Demultiplexing | int | Reads assigned to the sample after demultiplexing. |
| Reads_Mapped | int | Reads mapped to the reference sequence for the sample. |
| Coverage_Percentage | float | Base coverage of the reference sequence refers to how much of the reference sequence is covered by the reads. |
| Mean_Depth | float | Mean sequencing depth of the sample. |
| Mean_Base_Quality | float | Average sequencing quality of the bases. |
| Mean_Mapping_Quality | float | Average mapping quality of the reads. |
| ref_DNA_Length | int | Length of the reference sequence. |
| DNA_Identity_{*} | fraction | The number of bases that are identical to the reference sequence at each step. |
| DNA_Identity_pct_{*} | float | The percentage of bases that match the reference sequence at each step. |
| AssembledDNA_length_{*} | int | The length of the assembled DNA at each step. |
| ntCorrect_{*} | bool | Indicates if the nucleotide sequence is identical to the reference sequence at each step. |
| CompleteAssembled | bool | Indicates whether the assembly is complete, with matched bases exceeding 95%. |
| Pro_Identity_{*} | fraction | The number of matched amino acids in the sequence at each step. |
| Pro_Identity_pct_{*} | float | The percentage of matched amino acids in the sequence at each step. |
| RefORFlength_{*} | int | The length of ORF of the reference sequence (only valid when the input sequence is CDS). |
| ORFlength_{*} | int | The length of ORF of assembled sequence at each step (only valid when the input sequence is CDS). |
| aaCorrect_{*} | bool | Indicate if the amino acid sequence is identical to the reference sequence at each step (only valid when the input sequence is CDS). |
| aaSameLength_{*} | bool | Indicate if the length of amino acid sequence is the same as the reference sequence (only valid when the input sequence is CDS). |
| Variant_{*} | string | Type of variants at each step. |
| missenseMutation_{*} | bool | Indicates if the sample has missense mutation in each step (only valid when the input sequence is CDS). |
| aaMutation_medaka_1rd | bool | Indicates if the sample retains the same amino acid sequence (only valid when the input sequence is CDS). |
| Attention_LowDepth | bool | Indicates if the sample has low sequencing depth. |
| Alignment_{*} | link | Hyperlinks to the pairwise alignments of the assembled sequences generated by samtools and medaka polishing against the reference sequence. |
| Alignment_{*}_all_steps | link | Hyperlinks to the pairwise alignments of the assembled sequences generated at all steps against the reference sequence. |
| Alignment_{*}_all_clones | link | Hyperlinks to the pairwise alignments of the assembled sequences generated in the end (after medaka polishing) of all clones against the reference sequence. |
| AssembledDNALength | int | The length of the final assembled DNA. |
| ConsensusSequence | string | Consensus sequence generated in the final step. |
| ConsensusSequence_sam | string | Consensus sequence generated by samtools. |
| ReferenceSequence | string | Sequence of the reference. |
| CheckINDEL | bool | Indicates if the sample have the false INDELs. |
| Attention_PossibleHetSites | bool | Indicates if the sample has possible heterozygous sites. |
| Assembly_Status | string | The status of the assembly: it can be error-free (error-free), partially assembled (unassembled), or contain SNPs or INDELs (snp_indel). |  

[Back to README](https://github.com/ZenanXing/Construct-Validation-for-IGGYPOPseq?tab=readme-ov-file#output)

### Description of exported files

The results of all steps from the pipeline will be stored in a folder named "Analysis_Results" in your output directory and structured as follows:

```
(i)                             (ii)                              (iii)
Inputfiles                      Demultiplexing                    Alignment
├── references                  ├── final                         ├── set1_1.sam
│   ├── ref_for_set1.fasta      │   ├── set1_1.fastq              ├── set1_1.bam
│   └── ...                     │   ├── set1_1_filtered.fastq     ├── set1_1.sorted.bam
├── config_sample.txt           │   └── ...                       ├── set1_1.sorted.bam.bai
├── primer_index_info.tsv       ├── config.txt                    └── ...
├── references_info.tsv         └── e_n_E_Combination.tsv
├── IndexCombination.txt
└── references.tsv


(iv)                                        (v)                                           (vii)
Assembly                                    VariantCalling                                VariantIdentification
├── bedtools                                ├── beforePolishing                           ├── ReferenceORFs
│   ├── sam                                 │   ├── sample_1_bP.vcf                       │   ├── sample_1_bP.vcf
│   │   ├── set1_1_sam.fasta                │   └── ...                                   │   └── ...
│   │   ├── set1_1_sam_f.fasta              └── afterPolishing                            └── AssembledORFs
│   │   └── ...                                 ├── racon                                     ├── selectedORF
│   └── bed                                     │   ├── 1rd                                   │   ├── set1_1
│       ├── set1_1_draft.fasta                  │   │   ├── set1_1_racon_1rd.all.vcf          │   │   ├── set1_draft_orf.fasta
│       ├── set1_1_draft_f.fasta                │   │   └── ...                               │   │   ├── set1_1_racon_1rd_orf.fasta
│       └── ...                                 │   ├── 2rd                                   │   │   ├── set1_1_racon_3rd_orf.fasta
├── racon                                       │   └── 3rd                                   │   │   ├── set1_1_medaka_1rd_orf.fasta
│   ├── 1rd                                     └── medaka_1rd                                │   │   ├── set1_1_racon_2rd_orf.fasta
│   │   ├── set1_1_racon_1rd.fasta                  ├── set1_1_medaka_1rd.all.vcf             │   │   └── set1_1_sam_orf.fasta
│   │   ├── set1_1_racon_1rd_f.fasta                └── ...                                   │   └── ...
│   │   └── ...                             (vi)                                              ├── clones
│   ├── 2rd                                 PairwiseAlignment                                 │   ├── set1_comb.fasta
│   ├── 3rd                                 ├── combined_fasta                                │   └── ...
│   └── sam                                 │   ├── set1_1_comb.fasta                         ├── combinedORFs
└── medaka                                  │   ├── set1_1_comb_cleaned.fasta                 └── allORFs
    └── 1rd                                 │   └── ...                                 
        ├── set1_1_medaka_1rd.fasta         └── clones                                  
        ├── set1_1_medaka_1rd_f.fasta           ├── set1_comb.fasta                     
        └── ...                                 └── ...                                 
                                          

(viii)
OutputFiles
├── Summary.xlsx                            ├── Demultiplexing_ReadsSummary.tsv
├── Report.html                             ├── Alignment.tsv
├── ntAlignment                             ├── Assembled_bP.tsv
│   ├── set1_1_nt_align_muscle_f.html       ├── Racon_1rdP.tsv
│   ├── set1_1_nt_align_muscle.html         ├── Racon_2rdP.tsv
│   ├── set1_nt_align_muscle_clones.html    ├── Racon_3rdP.tsv
│   ├── set1_1_nt_align_emboss.txt          ├── Medaka_1rdP.tsv
│   └── ...                                 ├── VariantCalling_racon_1rd.tsv
├── aaAlignment                             ├── VariantCalling_racon_2rd.tsv
│   ├── set1_aa_align_muscle_clones.html    ├── VariantCalling_racon_3rd.tsv
│   ├── set1_1_aa_align_muscle.html         ├── VariantCalling_medaka_1rd.tsv
│   ├── set1_1_aa_align_muscle_f.html       ├── ntAlignment.tsv
│   ├── set1_1_aa_align_emboss.txt          ├── aaAlignment.tsv
│   └── ...                                 ├── Ref_ORF_length.tsv
                                            └── ORF_length.tsv

```

Output files may be aggregated including information after each step for all samples or provided per sample. Per-sample files will be prefixed with respective aliases and represented below as {{ alias }}.

| Step | Type | File path | Description | Per sample or aggregated |
|------|------|-----------|-------------|--------------------------|
| Demultiplexing | Sorted reads | ./Demultiplexing/final/{{ alias }}.fastq | Reads demultiplexed to each sample. | per-sample |
| Demultiplexing | Filtered reads | ./Demultiplexing/final/{{ alias }}_filtered.fastq | Filtered reads of each sample. | per-sample |
| Demultiplexing | Combination of e and E | ./Demultiplexing/e_n_E_Combination.tsv | Reads filtered and mapped using different combinations of parameter e and E in the demultiplexing. | aggregated |
| Demultiplexing | Summary | ./OutputFiles/Demultiplexing_ReadsSummary.tsv | Stats report of read count after filtering. | aggregated |
| Mapping | Alignment SAM file | ./Alignment/{{ alias }}.sam | Sam files generated after mapping by `minimap2`. | per-sample |
| Mapping | Summary | ./OutputFiles/Alignment.tsv | Stats report after mapping (e.g. Mapped Reads, Coverage, and Mean Depth, et. al.). | aggregated |
| Assembly | Consensus Sequence | ./Assembly/bedtools/sam/{{ alias }}_sam_f.fasta | Consensus Sequences generated by `samtools`. | per-sample |
| Assembly | Consensus Sequence | ./Assembly/bedtools/bed/{{ alias }}_draft_f.fasta | Sequences generated by `bedtools` after applying variant called by `bcftools`. | per-sample |
| Assembly | Consensus Sequence | ./Assembly/racon/{1rd/2rd/3rd}/{{ alias }}_racon_1rd_f.fasta | Sequence generated after first round of Racon polishing. The sequences generated after second and third rounds of polishing are stored in the folders named `2rd` and `3rd`, respectively. | per-sample |
| Assembly | Consensus Sequence | ./Assembly/medaka/1rd/{{ alias }}_medaka_1rd_f.fasta | Sequence generated after one round of Medaka polishing. This sequence is considered as the final consensus sequence in the analysis. | per-sample |
| Assembly | Summary | ./OutputFiles/Assembled_bP.tsv | Stats report of the draft consensus sequence after assembly, including details such as the sequence length. | aggregated |
| Assembly | Summary | ./OutputFiles/Racon_{1/2/3}rdP.tsv | Stats report of the consensus sequence generated after different rounds of `Racon` polishing. | aggregated |
| Assembly | Summary | ./OutputFiles/Medaka_1rdP.tsv | Stats report of the consensus sequence generated after first round of `Medaka` polishing. | aggregated |
| Variant Calling | Variants BCF files | ./VariantCalling/beforePolishing/{{ alias }}_bP.vcf | BCF file with variants found in the draft sequence. | per-sample |
| Variant Calling | Variants BCF files | ./VariantCalling/afterPolishing/racon/{1/2/3}rd/{{ alias }}_racon_{1/2/3}rd.all.vcf | BCF file with all the variants found after different rounds of `Racon` polishing. | per-sample |
| Variant Calling | Variants BCF files | ./VariantCalling/afterPolishing/medaka_1rd/{{ alias }}_medaka_1rd.all.vcf | BCF file with all the variants found after `Medaka` polishing. | per-sample |
| Variant Calling | Summary | ./OutputFiles/VariantCalling_racon_{1/2/3}rd.tsv | Stats report of variant calling after different rounds of `Racon` polishing. | aggregated |
| Variant Calling | Summary | ./OutputFiles/VariantCalling_medaka_1rd.tsv | Stats report of variant calling after `Medaka` polishing. | aggregated |
| Pairwise Alignment | Alignment HTML file | ./ntAlignment/{{ alias }}_nt_align_muscle.html | Pairwise alignment of nucleotide sequences generated after all steps against the reference sequence, using `muscle3`. | per-sample |
| Pairwise Alignment | Alignment HTML file | ./ntAlignment/{{ alias }}_nt_align_muscle_f.html | Pairwise alignment of draft nucleotide sequence (generated by `samtools`) and final nucleotide sequence (after `Medaka` polishing) against the reference sequence, using `muscle3`. | per-sample |
| Pairwise Alignment | Alignment HTML file | ./ntAlignment/{{ Set Name }}_nt_align_muscle_clones.html | Pairwise alignment of final nucleotide sequence (after `Medaka` polishing) of all the clones against the reference sequence, using `muscle3`. | per-sample |
| Pairwise Alignment | Alignment txt file | ./ntAlignment/{{ alias }}_nt_align_emboss.txt | Pairwise alignment of nucleotide sequences generated after all steps against the reference sequence, using `emboss`. | per-sample |
| Pairwise Alignment | Alignment HTML file | ./aaAlignment/{{ alias }}_aa_align_muscle.html | Pairwise alignment of amino acid sequences generated after all steps against the reference sequence, using `muscle3`. Only available when `--amino_acid_seq_validation` is set to `TRUE`. | per-sample |
| Pairwise Alignment | Alignment HTML file | ./aaAlignment/{{ alias }}_aa_align_muscle_f.html | Pairwise alignment of draft amino acid sequence (generated by `samtools`) and final amino acid sequence (after `Medaka` polishing) against the reference sequence, using `muscle3`. Only available when `--amino_acid_seq_validation` is set to `TRUE`. | per-sample |
| Pairwise Alignment | Alignment HTML file | ./aaAlignment/{{ Set Name }}_aa_align_muscle_clones.html | Pairwise alignment of final amino acid sequence (after `Medaka` polishing) of all the clones against the reference sequence, using `muscle3`. Only available when `--amino_acid_seq_validation` is set to `TRUE`. | per-sample |
| Pairwise Alignment | Alignment txt file | ./aaAlignment/{{ alias }}_aa_align_emboss.txt | Pairwise alignment of amino acid sequences generated after all steps against the reference sequence, using `emboss`. Only available when `--amino_acid_seq_validation` is set to `TRUE`. | per-sample |
| Pairwise Alignment | Summary | ./OutputFiles/ntAlignment.tsv | Stats report of pairwise alignment of nucleotide sequences after all steps. | aggregated |
| Pairwise Alignment | Summary | ./OutputFiles/aaAlignment.tsv | Stats report of pairwise alignment of amino acid sequences after all steps. | aggregated |
| Summary | Report | ./OutputFiles/Summary.xlsx | Stats of results obtained from all steps across all samples. | aggregated |
| Summary | Report | ./OutputFiles/Report.html | Report for all samples. | aggregated |  

[Back to README](https://github.com/ZenanXing/Construct-Validation-for-IGGYPOPseq?tab=readme-ov-file#output)
